examples/tcp-sockets/tcp-sockets.go

```diff
@@ -0,0 +1,173 @@
+	// address and port. The first argument "tcp" specifies
+	// the network protocol, and ":8080" means listen on
+	// all available network interfaces on port 8080.
+	// It returns a `net.Listener` interface that can be
+	// used to accept incoming connections.
+	listener, err := net.Listen("tcp", ":8080")
+	if err != nil {
+		panic(err)
+	}
+	// Always close the listener when done to free up
+	// the port and other resources.
+	defer listener.Close()
+
+	fmt.Println("TCP Server listening on :8080")
+
+	// Accept connections in a loop. Each connection
+	// represents a client connecting to the server.
+	// This is a blocking call - the server will wait
+	// here until a client connects.
+	for {
+		// `Accept` blocks until a new connection is made.
+		// It returns a `net.Conn` interface representing
+		// the connection, which provides methods for
+		// reading from and writing to the client.
+		conn, err := listener.Accept()
+		if err != nil {
+			// If there's an error accepting (e.g., the
+			// listener was closed), log it and continue
+			// accepting other connections.
+			fmt.Println("Error accepting connection:", err)
+			continue
+		}
+
+		// Handle each connection in a goroutine to allow
+		// the server to accept multiple clients concurrently.
+		// Without goroutines, the server would only be able
+		// to handle one client at a time, blocking others
+		// from connecting.
+		go handleTCPConnection(conn)
+	}
+}
+
+// handleTCPConnection processes a single client connection.
+// It reads data from the client and sends a response back.
+func handleTCPConnection(conn net.Conn) {
+	// Ensure the connection is closed when this function
+	// returns, freeing up resources. This is important
+	// because TCP connections consume system resources.
+	defer conn.Close()
+
+	// Read data from the connection using a buffered reader.
+	// `bufio.NewReader` wraps the connection with buffering,
+	// which improves performance by reducing the number of
+	// system calls. `ReadString('\n')` reads until it
+	// encounters a newline character, which is common for
+	// line-based protocols.
+	reader := bufio.NewReader(conn)
+	message, err := reader.ReadString('\n')
+	if err != nil {
+		// Handle read errors (e.g., client disconnected,
+		// network issues). We return instead of panicking
+		// so the server can continue handling other clients.
+		fmt.Println("Error reading:", err)
+		return
+	}
+
+	fmt.Printf("Received: %s", message)
+
+	// Send a response back to the client. TCP guarantees
+	// that data sent will arrive at the other end, though
+	// not necessarily immediately (it may be buffered).
+	// We append a newline character so the client can use
+	// `ReadString('\n')` to read the complete response.
+	response := fmt.Sprintf("Server received: %s", message)
+	_, err = conn.Write([]byte(response + "\n"))
+	if err != nil {
+		fmt.Println("Error writing:", err)
+	}
+}
+
+// The TCP client connects to a server and sends data.
```

Comment:
I feel like this sample is too busy, seems like much larger than other samples.

Maybe the client and server should be separated? `netcat` can be used for testing, like `curl` is for the HTTP server example